lib/mix/lib/mix/tasks/xref.ex

```diff
@@ -283,6 +320,85 @@ defmodule Mix.Tasks.Xref do
     Mix.raise message
   end
 
+  ## Graph helpers
+
+  defp excluded(opts) do
+    Keyword.get_values(opts, :exclude)
+    |> Enum.map(&{&1, nil})
+  end
+
+  defp module_references() do
+    source_modules =
+      for manifest <- E.manifests(),
+          manifest_data = read_manifest(manifest, ""),
+          module(module: module, source: source) <- manifest_data,
+          source = Enum.find(manifest_data, &match?(source(source: ^source), &1)),
+          do: {source, module}
+
+    source_modules = Enum.group_by(source_modules, &elem(&1, 0), &elem(&1, 1))
+
+    all_modules =
+      Enum.reduce source_modules, MapSet.new(), fn {_, modules}, acc ->
+        Enum.reduce(modules, acc, &MapSet.put(&2, &1))
+      end
+
+    Enum.reduce source_modules, [], fn {source, modules}, acc ->
+      source(runtime_references: r, compile_references: c) = source
+      references =
+        r ++ c ++ modules
+        |> MapSet.new()
+        |> MapSet.intersection(all_modules)
+
+      Enum.reduce modules, acc, fn module, acc ->
+        module_references = MapSet.delete(references, module)
+        [{{inspect(module), nil}, Enum.map(module_references, &{inspect(&1), nil})} | acc]
+      end
+    end
+  end
+
+  defp write_graph(module_references, excluded, opts) do
+    {root, module_references} =
+      case {opts[:source], opts[:sink]} do
+        {nil, nil} -> {{nil, nil}, module_references}
+        {source, nil} -> {{source, nil}, module_references}
+        {nil, sink} -> {{sink, nil}, invert_references(module_references)}
+        {_, _} -> Mix.raise "mix xref graph expects only one of --source and --sink"
+      end
+
+    callback =
+      fn
+        {nil, nil} ->
+          {{nil, nil}, Enum.map(module_references, &elem(&1, 0)) -- excluded}
+
+        {module, nil} ->
+          {_, children} = List.keyfind(module_references, {module, nil}, 0, {{module, nil}, []})
+          {{module, nil}, children -- excluded}
+      end
+
+    if opts[:format] == "dot" do
+      Mix.Utils.write_dot_graph!("xref_graph.dot", "xref graph",
+                                 root, callback, opts)
+      Mix.shell.info """
+        Generated "xref_graph.dot" in the current directory. To generate a PNG:
```

Comment:
Nitpick: this heredoc should be indented on the same level as the `Mix.shell.info` call :)